src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/BoundPropertyManager.cs

```diff
@@ -0,0 +1,246 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    public readonly struct BoundPropertyManager
+    {
+        public static readonly BoundPropertySource[] DefaultAllowedSources = new[]
+        {
+            BoundPropertySource.TempData,
+            BoundPropertySource.ViewData,
+        };
+
+        private BoundPropertyManager(IReadOnlyList<PropertyItem> propertyItems)
+        {
+            PropertyItems = propertyItems;
+        }
+
+        // Internal for unit testing
+        internal IReadOnlyList<PropertyItem> PropertyItems { get; }
+
+        public static BoundPropertyManager Create(
+            MvcViewOptions options,
+            Type type,
+            IReadOnlyList<BoundPropertySource> allowedSources = null,
+            bool requireViewDataPropertySetters = false)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (type == null)
+            {
+                throw new ArgumentNullException(nameof(type));
+            }
+
+            allowedSources = allowedSources ?? DefaultAllowedSources;
+
+            var tempDataPrefix = options.SuppressTempDataPropertyPrefix ?
+                string.Empty :
+                "TempDataProperty-";
+
+            var properties = GetProperties(type, allowedSources, tempDataPrefix, requireViewDataPropertySetters);
+            return new BoundPropertyManager(properties ?? Array.Empty<PropertyItem>());
+        }
+
+        public void Populate(object instance, BoundPropertyContext context)
+        {
+            if (instance == null)
+            {
+                throw new ArgumentNullException(nameof(instance));
+            }
```

Comment:
Null-check `context` too.

Reply:
It's a value type (same response for the other null check comments)